Langsmith "Add to Dataset" Error with Pydantic Formatting Output

Hi all,

Using Langsmith to manage and tests my LLM program, I use datasets since a while to benchmark the quality of my prompts.
Recently I tried to add a new example to my dataset by using the button `Add to Dataset` while looking at the ChatOpenAI run from a past execution.
I used to do that and it worked perfectly fine, but this new example fails to run, this might be due to the use of a Pydantic output formatting even tho LangSmith really looks like it can manage pretty well Pydantic output in datasets testing.
I provide below the screenshots and the error.


![image](https://github.com/langchain-ai/langserve/assets/73235385/80c6430f-106e-4780-a8b1-60b49e3e9f00)


The output should be a Pydantic object:

{
  "description": "...",
  "grade": "good" or "bad"
}



Using the "Add to Dataset" button automatically creates an Example for a Chat Dataset like this:


![image](https://github.com/langchain-ai/langserve/assets/73235385/f26e83d2-9c92-4929-aed6-2aa031142492)

So the output is indeed as expected: the Pydantic format CodeReviewResult.
I then have this Test Run:


```python
from datetime import datetime

import langsmith
from langchain import smith

from langchain_openai import ChatOpenAI
from app.cobol_enhancer.common import MODEL_NAME

date = int(datetime.now().strftime('%Y%m%d%H%M%S'))
model = ChatOpenAI(temperature=0, model=MODEL_NAME, streaming=True)

# Define the evaluators to apply
eval_config = smith.RunEvalConfig(
    evaluators=[
        "cot_qa",
        smith.RunEvalConfig.LabeledCriteria("conciseness"),
        smith.RunEvalConfig.LabeledCriteria("coherence")
    ],
    custom_evaluators=[],
    eval_llm=ChatOpenAI(model="gpt-4", temperature=0)
)

def test_critic_generation_2():
    client = langsmith.Client()
    chain_results = client.run_on_dataset(
        dataset_name="critic_generation_dataset_2",
        llm_or_chain_factory=model,
        evaluation=eval_config,
        project_name=f"critic_generation-{date}",
        concurrency_level=5,
        verbose=True,
    )
```


And I get the following error:
```bash
 test summary info =======================================================
==============================================================
=========================
FAILED tests/test_datasets.py::test_critic_generation - langchain.smith.evaluation.runner_utils.InputFormatError:
Example inputs do not match language model input format.
Expected a dictionary with messages or a single prompt.
Got: {'input': [{'data': {'content': 'You are an expert in code analysis with a focus on COBOL. Examine the original code, the versi...

```


After some research and by manually creating the example in a Chat/LLM/Key-Value dataset I found out that I can only get it to work by not specifying the Pydantic function into LangSmith and by only specifying the test output as a String:

![image](https://github.com/langchain-ai/langserve/assets/73235385/4caf429a-08bb-45b3-b326-68903a46dfd4)

Am I missing something to use Pydantic with dataset example? I couldn't find any info or help on other issues.
Also I'm sorry if this is not exactly the place to post for Langsmith question but I couldn't find any specific chat forum or smth else.

Thanks for your help and your incredible work with Langchain.

--- a/libs/langchain/langchain/smith/evaluation/runner_utils.py
+++ b/libs/langchain/langchain/smith/evaluation/runner_utils.py
@@ -319,7 +319,7 @@ def _get_messages(inputs: Dict[str, Any]) -> dict:
     else:
         raise InputFormatError(
             f"Chat Run expects single List[dict] or List[List[dict]] 'messages'"
-            f" input. Got {len(raw_messages)} messages from inputs {inputs}"
+            f" input. Got {inputs}"
         )
 
 

--- a/libs/langchain/langchain/smith/evaluation/runner_utils.py
+++ b/libs/langchain/langchain/smith/evaluation/runner_utils.py
@@ -280,7 +280,11 @@ def _get_prompt(inputs: Dict[str, Any]) -> str:
         )
 
 
-def _get_messages(inputs: Dict[str, Any]) -> List[BaseMessage]:
+class ChatModelInput(TypedDict):
+    messages: List[BaseMessage]
+
+
+def _get_messages(inputs: Dict[str, Any]) -> dict:
     """Get Chat Messages from inputs.
 
     Args:
@@ -293,31 +297,25 @@ def _get_messages(inputs: Dict[str, Any]) -> List[BaseMessage]:
     """
     if not inputs:
         raise InputFormatError("Inputs should not be empty.")
-
+    input_copy = inputs.copy()
     if "messages" in inputs:
-        single_input = inputs["messages"]
+        input_copy["input"] = input_copy.pop("messages")
     elif len(inputs) == 1:
-        single_input = next(iter(inputs.values()))
-    else:
-        raise InputFormatError(
-            f"Chat Run expects 'messages' in inputs when example has multiple"
-            f" input keys. Got {inputs}"
-        )
-    if isinstance(single_input, list) and all(
-        isinstance(i, dict) for i in single_input
-    ):
-        raw_messages = [single_input]
-    elif isinstance(single_input, list) and all(
-        isinstance(i, list) for i in single_input
-    ):
-        raw_messages = single_input
-    else:
-        raise InputFormatError(
-            f"Chat Run expects List[dict] or List[List[dict]] values for"
-            f" 'messages' key input. Got {inputs}"
-        )
-    if len(raw_messages) == 1:
-        return messages_from_dict(raw_messages[0])
+        input_copy["input"] = next(iter(inputs.values()))
+    if "input" in input_copy:
+        raw_messages = input_copy["input"]
+        if isinstance(raw_messages, list) and all(
+            isinstance(i, dict) for i in raw_messages
+        ):
+            raw_messages = [raw_messages]
+        if len(raw_messages) == 1:
+            input_copy["input"] = messages_from_dict(raw_messages[0])
+        else:
+            raise InputFormatError(
+                "Batch messages not supported. Please provide a"
+                " single list of messages."
+            )
+        return input_copy
     else:
         raise InputFormatError(
             f"Chat Run expects single List[dict] or List[List[dict]] 'messages'"
@@ -711,9 +709,9 @@ async def _arun_llm(
                 ),
             )
         except InputFormatError:
-            messages = _get_messages(inputs)
+            llm_inputs = _get_messages(inputs)
             llm_output = await llm.ainvoke(
-                messages,
+                **llm_inputs,
                 config=RunnableConfig(
                     callbacks=callbacks, tags=tags or [], metadata=metadata or {}
                 ),
@@ -864,9 +862,9 @@ def _run_llm(
                 ),
             )
         except InputFormatError:
-            llm_messages = _get_messages(inputs)
+            llm_inputs = _get_messages(inputs)
             llm_output = llm.invoke(
-                llm_messages,
+                **llm_inputs,
                 config=RunnableConfig(callbacks=callbacks, metadata=metadata or {}),
             )
     return llm_output

